### PR TITLE
Remove createMemoryRouter from reference panel

### LINKED ISSUE #37489: Remove MemoryRouter from ReferencesPanel as it blocks React Router v6 migration
## Context
The @sourcegraph/frontend-platform team is now [migrating from Router v5 --> v6](https://github.com/sourcegraph/sourcegraph/issues/33834). This includes staring using `CompatRouter` first to wrap all sub-routes, and later replacing all `Router` and `Route` components with the new versions.

There was also a [crash on dotcom](https://sourcegraph.slack.com/archives/C01LTKUHRL3/p1655395878523209) related to this update.

## ReferencesPanel
However, one of @sourcegraph/code-intel's features, ReferencesPanel, is using MemoryRouter internally (inside the global Router) to support navigation via standard `Link`s without affecting app's global nav state. 

Unfortunately, rendering a MemoryRouter inside a Router in v6 doesn't work anymore, and [the following code](https://gist.github.com/oleggromov/ec7fe2c7ed3f4fb37b33f8a2ec1c25ec) will throw an exception.

### What could we do?
- Reimplement the same functionality without MemoryRouter (tricky and expensive to do)
- Get rid of this implementation of ReferencesPanel and think of something else
- ...

### PR DESCRIPTION
**Pull request by philipp-spiess**
Closes #37489 

Turns out removing the react router features from the references panel is relatively simple with the current changes:

1. References panel internal navigations were already using a custom state variable (https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/codeintel/ReferencesPanel.tsx?L320-328&subtree=true) so the custom memory state wasn't used for that.
2. The usage inside the Blob view seem _minimal_. After the latest changes from Olaf, we only use it to navigate to an external route which does not need a custom router anyway. And even if we were to change minds, we're already creating a [custom history](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/codeintel/ReferencesPanel.tsx?L706-710&subtree=true) inside the references panel only for the blob view. Since this is not using react router API that is at change with the v6 migration, we don't care about this either.

## Test plan

- Tested locally
- CI is green

## App preview:

- [Web](https://sg-web-ps-remove-nested-router-from-code.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.01% (-0.30 kb) | -0.00% (-0.40 kb) | -0.00% (-0.09 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ps/remove-nested-router-from-code-intel/compare-report.html;tab=live_object) for a full comparison between the commits e3de28c and 4338f05 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by philipp-spiess** [comment]
@valerybugakov Do you mind taking a look at the test that is failing or pairing with me to fix it? I've been banging my head against this for the last few hours and I didn't make any progress. As far as I understand, everything, `unstable_HistoryRouter` is used to set up a custom react router context that uses a history object so we can make assertions on where it redirects to but this fails somewhere within react router with no clear error message?

I have already tried upgrading `react-router-dom-v5-compat` but that failed in a different way:

```
 FAIL   web  client/web/src/codeintel/ReferencesPanel.test.tsx
  ● Test suite failed to run

    Cannot find module '@remix-run/router' from '../../node_modules/.pnpm/react-router-dom-v5-compat@6.7.0_xjeikfjhclro5pp2abrahotxli/node_modules/react-router-dom-v5-compat/dist/umd/react-router-dom-v5-compat.development.js'

    Require stack:
      /Users/philipp/dev/sourcegraph/node_modules/.pnpm/react-router-dom-v5-compat@6.7.0_xjeikfjhclro5pp2abrahotxli/node_modules/react-router-dom-v5-compat/dist/umd/react-router-dom-v5-compat.development.js
      /Users/philipp/dev/sourcegraph/node_modules/.pnpm/react-router-dom-v5-compat@6.7.0_xjeikfjhclro5pp2abrahotxli/node_modules/react-router-dom-v5-compat/dist/main.js
      src/codeintel/ReferencesPanel.test.tsx

      at Resolver._throwModNotFoundError (../../node_modules/.pnpm/jest-resolve@28.1.3/node_modules/jest-resolve/build/resolver.js:491:11)
      at ../../node_modules/.pnpm/react-router-dom-v5-compat@6.7.0_xjeikfjhclro5pp2abrahotxli/node_modules/react-router-dom-v5-compat/dist/umd/react-router-dom-v5-compat.development.js:12:126
      at Object.<anonymous> (../../node_modules/.pnpm/react-router-dom-v5-compat@6.7.0_xjeikfjhclro5pp2abrahotxli/node_modules/react-router-dom-v5-compat/dist/umd/react-router-dom-v5-compat.development.js:15:3)
```

I wasn't able to find where on the file system the file is persisted so I can do console.log debugging either. Very frustrating.

I also locally can't get `pnpm storybook` to work because of another pnpm related error. What's strange is that others (I asked Vova) do not have these errors:

```
α sourcegraph (ps/remove-nested-router-from-code-intel) pnpm storybook

> @ storybook /Users/philipp/dev/sourcegraph
> pnpm --filter @sourcegraph/storybook run start


> @sourcegraph/storybook@0.0.1 start /Users/philipp/dev/sourcegraph/client/storybook
> TS_NODE_TRANSPILE_ONLY=true start-storybook -p 9001 -c ./src

node:internal/modules/cjs/loader:988
  throw err;
  ^

Error: Cannot find module '/Users/philipp/dev/sourcegraph/node_modules/.pnpm/@storybook+react@6.5.14_3rki7qa32pb6vtxta3ajy52kte/node_modules/@storybook/react/bin/index.js'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:985:15)
    at Function.Module._load (node:internal/modules/cjs/loader:833:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:22:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
/Users/philipp/dev/sourcegraph/client/storybook:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @sourcegraph/storybook@0.0.1 start: `TS_NODE_TRANSPILE_ONLY=true start-storybook -p 9001 -c ./src`
Exit status 1
 ELIFECYCLE  Command failed with exit code 1.

```

File: package.json [hunk 1]
```diff
@@ -487,6 +487,15 @@
...
   "packageManager": "pnpm@7.24.2",
+  "pnpm": {
+    "packageExtensions": {
+      "react-router-dom-v5-compat": {
+        "dependencies": {
+          "@remix-run/router": "*"
+        }
+      }
+    }
+  },
```
Comment (valerybugakov) [inline package.json:498]: Based on https://github.com/remix-run/react-router/issues/9243#issuecomment-1246456092

File: client/web/src/codeintel/ReferencesPanel.mocks.ts [hunk 2]
```diff
@@ -304,6 +300,7 @@ const HIGHLIGHTED_FILE_MOCK = {
             commit: {
                 id: 'R2l0Q29tbWl0OnsiciI6IlVtVndiM05wZEc5eWVUb3lNRFE9IiwiYyI6IjlkMWYzNTNhMjg1YjMwOTRiYzMzYmRhZTI3N2ExOWFlZGFiZThiNzEifQ==',
                 blob: {
+                    content: 'content',
```
Comment (valerybugakov) [inline client/web/src/codeintel/ReferencesPanel.mocks.ts:303]: Fixes a non-critical error printed by the Apollo Client cache.

**Comment by valerybugakov** [comment]
@philipp-spiess, it looks like our CI is happy with the last commit! 

Re: storybook
Let's discuss it on our today's sync call. I verified that it works for me locally, and the only suggestion I have from the top of my head is: `rm -rf node_modules ./client/*/node_modules && pnpm i`, which I think you already tried. 🙂

**Comment by philipp-spiess** [comment]
@valerybugakov I haven't tried deleting the node_modules files of _all_ dependencies but that helped, yay <3

**Review by valerybugakov — APPROVED** [review]
Great stuff. We're one step closer to removing the react-router-compat layer!

**Pull request merged**